velox/exec/GroupingSet.cpp

```diff
@@ -1081,6 +1095,44 @@ bool GroupingSet::mergeNext(
   VELOX_UNREACHABLE();
 }
 
+bool GroupingSet::mergeNextForDistinct(
+    int32_t maxOutputRows,
+    const RowVectorPtr& result) {
+  VELOX_CHECK(isDistinct());
+
+  bool nextKeyIsEqual{false};
+  bool newDistinct{true};
+  int32_t numOutputRows{0};
+  for (;;) {
+    const auto next = merge_->nextWithEquals();
+    auto* stream = next.first;
+    if (stream == nullptr) {
+      break;
+    }
+    if (stream->id() == 0) {
+      // A spilled row is not distinct if it is contained in the first stream
+      // which has already been output as distinct before we triggered the
+      // spilling.
+      newDistinct = false;
+    }
+    if (next.second) {
+      stream->pop();
+      continue;
+    }
+    if (newDistinct) {
```

Comment:
Why not just if (stream->id() != 0)

or make 'newDistinct' const local variable inside the loop?

Reply:
It might be better not to assume the tree of loser always return the row from the first stream with the same key?